pixy/consumer_test.go

```diff
@@ -0,0 +1,486 @@
+			"c": map[int32]none{4: nothing, 5: nothing},
+		})
+	c.Assert(resolveAssignments([]int32{6, 0, 3, 1, 2, 5, 4}, []string{"b", "c", "a"}),
+		DeepEquals, map[string]map[int32]none{
+			"a": map[int32]none{0: nothing, 1: nothing, 2: nothing},
+			"b": map[int32]none{3: nothing, 4: nothing},
+			"c": map[int32]none{5: nothing, 6: nothing},
+		})
+	c.Assert(resolveAssignments([]int32{6, 0, 3, 1, 2, 5, 4}, []string{"d", "b", "c", "a"}),
+		DeepEquals, map[string]map[int32]none{
+			"a": map[int32]none{0: nothing, 1: nothing},
+			"b": map[int32]none{2: nothing, 3: nothing},
+			"c": map[int32]none{4: nothing, 5: nothing},
+			"d": map[int32]none{6: nothing},
+		})
+}
+
+// If a topic has the only partition then the consumer retrieve messages in the
+// order they were produced.
+func (s *SmartConsumerSuite) TestSinglePartitionTopic(c *C) {
+	// Given
+	ResetOffsets(c, "group-1", "test.1")
+	produced := GenMessages(c, "single", "test.1", map[string]int{"": 3})
+
+	sc, err := SpawnSmartConsumer(s.newConfig("consumer-1"))
+	c.Assert(err, IsNil)
+
+	// When/Then
+	consumed := s.consume(c, sc, "group-1", "test.1", 1)
+	s.assertMsg(c, consumed[""][0], produced[""][0])
+
+	sc.Stop()
+}
+
+// If we stop one consumer and start another, the new one picks up where the
+// previous one left off.
+func (s *SmartConsumerSuite) TestSequentialConsume(c *C) {
+	// Given
+	ResetOffsets(c, "group-1", "test.1")
+	produced := GenMessages(c, "sequencial", "test.1", map[string]int{"": 3})
+
+	config := s.newConfig("consumer-1")
+	sc1, err := SpawnSmartConsumer(config)
+	c.Assert(err, IsNil)
+	log.Infof("*** GIVEN 1")
+	consumed := s.consume(c, sc1, "group-1", "test.1", 2)
+	s.assertMsg(c, consumed[""][0], produced[""][0])
+	s.assertMsg(c, consumed[""][1], produced[""][1])
+
+	// When: one consumer stopped and another one takes its place.
+	log.Infof("*** WHEN")
+	sc1.Stop()
+	sc2, err := SpawnSmartConsumer(config)
+	c.Assert(err, IsNil)
+
+	// Then: the second message is consumed.
+	log.Infof("*** THEN")
+	consumed = s.consume(c, sc2, "group-1", "test.1", 1, consumed)
+	s.assertMsg(c, consumed[""][2], produced[""][2])
+	sc2.Stop()
+}
+
+// If a topic has several partitions then they all are being consumed in random
+// order.
+func (s *SmartConsumerSuite) TestMultiPartitionTopic(c *C) {
+	// Given
+	ResetOffsets(c, "group-1", "test.4")
+	GenMessages(c, "multi", "test.4", map[string]int{"A": 10, "B": 10})
+
+	log.Infof("*** GIVEN 1")
+	sc, err := SpawnSmartConsumer(s.newConfig("consumer-1"))
+	c.Assert(err, IsNil)
+
+	// When: exactly one half of all produced events is consumed.
+	log.Infof("*** WHEN")
+	sc.Consume("group-1", "test.4")
+	// Allow some time to spin up consumers for all partitions.
+	time.Sleep(300 * time.Millisecond)
+	consumed := s.consume(c, sc, "group-1", "test.4", 5)
+
+	// Then: we have events from both partitions consumed.
+	log.Infof("*** THEN")
+	delta := len(consumed["A"]) - len(consumed["B"])
+	if delta > 5 || delta < -5 {
```

Comment:
If we consumed exactly 5 events, then this check will never pass. If we had 5 "A" events and 0 "B" events, delta is not greater than 5. Maybe you meant `>=` ? That check will have a flaky failure once every 32 runs even if it is perfectly random, though. To check that there is a roughly even distribution I think I would check more events and also check that len(consumed["A"]) is between 1/4 and 3/4 of the total number of events consumed.